### PR TITLE
Add support for taking screenshots on Android

### LINKED ISSUE #115: Fallback to screenshots doesn't work
Currently, marathon doesn't capture screenshots if video recording is not available on the device.

### PR DESCRIPTION
**Pull request by Malinskiy**
Screenshots are taken all the time independent of the test result

Fixes #115

### CODE REVIEW

File: vendor-android/src/main/kotlin/com/malinskiy/marathon/android/AndroidDeviceProvider.kt [hunk 1]
```diff
@@ -42,37 +45,68 @@ class AndroidDeviceProvider : DeviceProvider {
...
+
+            private fun waitForBoot(device: AndroidDevice): Boolean {
+                var booted = false
+                for (i in 1..30) {
+                    if (device.booted) {
+                        logger.debug { "Device ${device.serialNumber} booted!" }
+                        booted = true
+                        break
+                    } else {
+                        Thread.sleep(1000)
```
Comment (tagantroy) [inline vendor-android/src/main/kotlin/com/malinskiy/marathon/android/AndroidDeviceProvider.kt]: Can you replace Thread.sleep with `delay` and use suspend modifier?

File: core/src/main/kotlin/com/malinskiy/marathon/execution/device/DeviceActor.kt [hunk 2]
```diff
@@ -172,6 +173,10 @@ class DeviceActor(private val devicePoolId: DevicePoolId,
         job = async(context, parent = deviceJob) {
             try {
                 device.execute(configuration, devicePoolId, batch, result, progressReporter)
+            } catch (e: DeviceLostException) {
+                logger.error(e) { "Critical error during execution" }
+                returnBatch(batch)
+//                terminate()
```
Comment (tagantroy) [inline core/src/main/kotlin/com/malinskiy/marathon/execution/device/DeviceActor.kt]: Are you sure that it is correct?

Comment (Malinskiy) [reply core/src/main/kotlin/com/malinskiy/marathon/execution/device/DeviceActor.kt]: Yes, for starters I want to just do the same logic, but log that it's a critical error that should lead in the end to device disconnected

File: vendor-android/src/main/kotlin/com/malinskiy/marathon/android/AndroidDevice.kt [hunk 3]
```diff
@@ -92,22 +94,29 @@ class AndroidDevice(val ddmsDevice: IDevice) : Device {
...
-                         deferred: CompletableDeferred<TestBatchResults>,
-                         progressReporter: ProgressReporter) {
-        AndroidDeviceTestRunner(this@AndroidDevice).execute(configuration, devicePoolId, testBatch, deferred, progressReporter)
+    override suspend fun execute(configuration: Configuration,
+                                 devicePoolId: DevicePoolId,
+                                 testBatch: TestBatch,
+                                 deferred: CompletableDeferred<TestBatchResults>,
+                                 progressReporter: ProgressReporter) {
+
+        launch(deviceContext) {
```
Comment (tagantroy) [inline vendor-android/src/main/kotlin/com/malinskiy/marathon/android/AndroidDevice.kt]: Why do you need suspend here if you just launch new coroutine and you are not handling result

File: core/src/main/kotlin/com/malinskiy/marathon/execution/DevicePoolActor.kt [hunk 4]
```diff
@@ -56,15 +57,15 @@ class DevicePoolActor(private val poolId: DevicePoolId,
         devices.filter {
             !it.value.isClosedForSend
         }.forEach {
-            it.value.send(DeviceEvent.WakeUp)
+            it.value.safeSend(DeviceEvent.WakeUp)
```
Comment (tagantroy) [inline core/src/main/kotlin/com/malinskiy/marathon/execution/DevicePoolActor.kt]: If you use this safeSend then please delete devices.filter

File: core/src/main/kotlin/com/malinskiy/marathon/execution/DevicePoolActor.kt [hunk 5]
```diff
@@ -56,15 +57,15 @@ class DevicePoolActor(private val poolId: DevicePoolId,
...
         }
     }
 
     private suspend fun onQueueTerminated() {
         devices.filterValues {
             !it.isClosedForSend
         }.forEach {
-            it.value.send(DeviceEvent.Terminate)
+            it.value.safeSend(DeviceEvent.Terminate)
```
Comment (tagantroy) [inline core/src/main/kotlin/com/malinskiy/marathon/execution/DevicePoolActor.kt]: same

File: core/src/main/kotlin/com/malinskiy/marathon/execution/DevicePoolActor.kt [hunk 6]
```diff
@@ -84,7 +85,7 @@ class DevicePoolActor(private val poolId: DevicePoolId,
     private suspend fun executeBatch(device: Device, batch: TestBatch) {
         devices[device.serialNumber]?.run {
             if (!isClosedForSend) {
-                send(DeviceEvent.Execute(batch))
+                safeSend(DeviceEvent.Execute(batch))
```
Comment (tagantroy) [inline core/src/main/kotlin/com/malinskiy/marathon/execution/DevicePoolActor.kt]: same

File: vendor-android/src/main/kotlin/com/malinskiy/marathon/android/AndroidDevice.kt (hunk 3 above)
Comment (Malinskiy) [reply vendor-android/src/main/kotlin/com/malinskiy/marathon/android/AndroidDevice.kt]: I want to have a goal of converging the threading from blocking vendor modules to using coroutines inside vendor implementation also. Currently it's blocked by ddmlib and ssh but we'll get there later.

It still asynchronously waits for the other coroutine finish in both prepare and execute though here.

File: core/src/main/kotlin/com/malinskiy/marathon/execution/DevicePoolActor.kt (hunk 6 above)
Comment (Malinskiy) [reply core/src/main/kotlin/com/malinskiy/marathon/execution/DevicePoolActor.kt]: Fixed

File: core/src/main/kotlin/com/malinskiy/marathon/execution/DevicePoolActor.kt (hunk 5 above)
Comment (Malinskiy) [reply core/src/main/kotlin/com/malinskiy/marathon/execution/DevicePoolActor.kt]: Fixed

File: core/src/main/kotlin/com/malinskiy/marathon/execution/DevicePoolActor.kt (hunk 4 above)
Comment (Malinskiy) [reply core/src/main/kotlin/com/malinskiy/marathon/execution/DevicePoolActor.kt]: Fixed

**Review by tagantroy — APPROVED** [review]

**Pull request merged**